File: services/llm_engine.py
"""
LLMEngine: a singleton, provider-agnostic router.

Round 1 had the Gemini multi-model-discovery + NVIDIA Llama-3.3 fallback
logic, JSON string-slicing, and prompt construction all inline inside
app.py. This module encapsulates the "call some LLM and get back a
validated object" concern so nothing above this layer needs to know which
provider answered, or care about code fences / stray commas in the raw
text.

Usage:
    engine = LLMEngine.get_instance()
    engine.configure(gemini_api_key=..., nvidia_api_key=...)
    result: AnalysisResult = engine.generate_structured(prompt, AnalysisResult)
"""
from __future__ import annotations

import json
import re
import threading
from typing import Any, Dict, List, Optional, Type, TypeVar

import google.generativeai as genai
from openai import OpenAI
from pydantic import BaseModel, ValidationError

SchemaT = TypeVar("SchemaT", bound=BaseModel)

# Preference order when multiple Gemini models are available to the key.
GEMINI_MODEL_PRIORITY: List[str] = [
    "models/gemini-3.6-flash",
    "models/gemini-1.5-flash",
    "models/gemini-1.5-pro",
    "models/gemini-2.0-flash",
    "models/gemini-pro",
]

NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1"
NVIDIA_MODEL = "meta/llama-3.3-70b-instruct-v2"


class LLMEngineError(RuntimeError):
    """Raised whenever no provider could produce a usable, schema-valid response."""


class LLMEngine:
    """Process-wide singleton. Safe to call `LLMEngine.get_instance()` anywhere."""

    _instance: Optional["LLMEngine"] = None
    _instance_lock = threading.Lock()

    def __init__(self) -> None:
        # Guard against accidental direct instantiation creating a second
        # "singleton" -- always go through get_instance().
        self._gemini_api_key: Optional[str] = None
        self._nvidia_api_key: Optional[str] = None
        self._call_lock = threading.Lock()

    @classmethod
    def get_instance(cls) -> "LLMEngine":
        if cls._instance is None:
            with cls._instance_lock:
                if cls._instance is None:
                    cls._instance = cls()
        return cls._instance

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def configure(self, gemini_api_key: Optional[str], nvidia_api_key: Optional[str]) -> None:
        """(Re)inject provider keys. Safe to call every Streamlit rerun -- it's just
        an attribute swap, not a network call."""
        self._gemini_api_key = gemini_api_key or None
        self._nvidia_api_key = nvidia_api_key or None

    def is_configured(self) -> bool:
        return bool(self._gemini_api_key) or bool(self._nvidia_api_key)

    # ------------------------------------------------------------------
    # Provider calls
    # ------------------------------------------------------------------

    def _call_gemini(self, prompt: str) -> str:
        genai.configure(api_key=self._gemini_api_key)
        available_models = [
            m.name for m in genai.list_models()
            if "generateContent" in m.supported_generation_methods
        ]

        selected_model_name = next(
            (candidate for candidate in GEMINI_MODEL_PRIORITY if candidate in available_models),
            None,
        )
        if selected_model_name is None and available_models:
            selected_model_name = available_models[0]
        if selected_model_name is None:
            raise RuntimeError("No Gemini models supporting generateContent are available.")

        model = genai.GenerativeModel(selected_model_name)
        response = model.generate_content(prompt)
        text = (response.text or "").strip()
        if not text:
            raise RuntimeError("Gemini returned an empty response.")
        return text

    def _call_nvidia(self, prompt: str) -> str:
        client = OpenAI(base_url=NVIDIA_BASE_URL, api_key=self._nvidia_api_key)
        completion = client.chat.completions.create(
            model=NVIDIA_MODEL,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=1500,
            temperature=0.2,
        )
        text = (completion.choices[0].message.content or "").strip()
        if not text:
            raise RuntimeError("NVIDIA returned an empty response.")
        return text

    def generate_text(self, prompt: str) -> str:
        """Resilient router: try Gemini, then fall back to the NVIDIA Llama-3.3 endpoint."""
        errors: List[str] = []

        with self._call_lock:
            if self._gemini_api_key:
                try:
                    return self._call_gemini(prompt)
                except Exception as exc:  # noqa: BLE001 - deliberately broad, this is a fallback boundary
                    errors.append(f"Gemini Error: {exc}")

            if self._nvidia_api_key:
                try:
                    return self._call_nvidia(prompt)
                except Exception as exc:  # noqa: BLE001
                    errors.append(f"NVIDIA Error: {exc}")

        if errors:
            raise LLMEngineError(" | ".join(errors))
        raise LLMEngineError("No LLM provider is configured.")

    # ------------------------------------------------------------------
    # JSON extraction + auto-repair
    # ------------------------------------------------------------------

    @staticmethod
    def _strip_code_fences(text: str) -> str:
        text = text.strip()
        if text.startswith("```"):
            text = text.strip("`")
            if text.lower().startswith("json"):
                text = text[4:]
            text = text.strip()
        return text

    @staticmethod
    def _isolate_json_object(text: str) -> str:
        start = text.find("{")
        end = text.rfind("}")
        if start == -1 or end == -1 or end < start:
            raise LLMEngineError("No JSON object found in the LLM response.")
        return text[start:end + 1]

    @staticmethod
    def _repair_common_json_issues(json_str: str) -> str:
        """Best-effort fixes for the mistakes small/medium LLMs make most often.

        This is intentionally a heuristic pass, not a full JSON5 parser -- it
        targets the failure modes actually seen from these prompts:
        trailing commas, smart quotes, and stray control characters.
        """
        # Smart/curly quotes -> straight quotes.
        json_str = json_str.replace("\u201c", '"').replace("\u201d", '"')
        json_str = json_str.replace("\u2018", "'").replace("\u2019", "'")
        # Trailing commas before a closing brace/bracket.
        json_str = re.sub(r",\s*([}\]])", r"\1", json_str)
        # Raw newlines/tabs inside the object (outside of what json expects).
        json_str = json_str.replace("\r", " ")
        return json_str

    @classmethod
    def extract_json(cls, raw_text: str) -> Dict[str, Any]:
        """Strip code fences, isolate the JSON object, and parse it -- repairing
        common formatting mistakes if the first parse attempt fails."""
        if not raw_text:
            raise LLMEngineError("Empty response from LLM provider.")

        text = cls._strip_code_fences(raw_text)
        json_slice = cls._isolate_json_object(text)

        try:
            return json.loads(json_slice)
        except json.JSONDecodeError:
            repaired = cls._repair_common_json_issues(json_slice)
            try:
                return json.loads(repaired)
            except json.JSONDecodeError as exc:
                raise LLMEngineError(f"Failed to parse JSON from LLM response: {exc}") from exc

    # ------------------------------------------------------------------
    # High-level structured entry point
    # ------------------------------------------------------------------

    def generate_structured(self, prompt: str, schema: Type[SchemaT]) -> SchemaT:
        """Call the resilient router, extract/repair JSON, then validate against
        `schema`. Raises LLMEngineError on any failure in the chain -- callers
        should catch this once and decide on a fallback (e.g. the legacy
        keyword-matching analyzer)."""
        raw_text = self.generate_text(prompt)
        data = self.extract_json(raw_text)
        try:
            return schema.model_validate(data)
        except ValidationError as exc:
            raise LLMEngineError(f"LLM output failed schema validation: {exc}") from exc

File: src/services/llmEngine.js
/**
 * LLMEngine: a singleton, provider-agnostic router.
 *
 * Direct port of services/llm_engine.py. Tries Gemini first, then falls
 * back to NVIDIA NIM models. Both providers are called over plain HTTPS via
 * axios instead of the Python SDKs.
 */
const axios = require("axios");

const GEMINI_MODEL_PRIORITY = [
  "models/gemini-2.0-flash",
  "models/gemini-1.5-flash",
  "models/gemini-1.5-pro",
  "models/gemini-pro",
];

const GEMINI_API_BASE = "https://generativelanguage.googleapis.com/v1beta";
const NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1";

// Ordered priority list for NVIDIA NIM models.
const NVIDIA_CANDIDATE_MODELS = [
  "google/diffusiongemma-26b-a4b-it",
  "moonshotai/kimi-k3",
  "meta/llama-3.3-70b-instruct",
  "mistralai/mistral-nemo-12b-instruct",
];

class LLMEngineError extends Error {}

class LLMEngine {
  constructor() {
    this._geminiApiKey = null;
    this._nvidiaApiKey = null;
  }

  static getInstance() {
    if (!LLMEngine._instance) {
      LLMEngine._instance = new LLMEngine();
    }
    return LLMEngine._instance;
  }

  configure({ geminiApiKey, nvidiaApiKey }) {
    this._geminiApiKey = geminiApiKey || null;
    this._nvidiaApiKey = nvidiaApiKey || null;
  }

  isConfigured() {
    return Boolean(this._geminiApiKey) || Boolean(this._nvidiaApiKey);
  }

  async _callGemini(prompt) {
    const listResp = await axios.get(`${GEMINI_API_BASE}/models`, {
      params: { key: this._geminiApiKey },
      timeout: 15000,
    });
    const availableModels = (listResp.data.models || [])
      .filter((m) => (m.supportedGenerationMethods || []).includes("generateContent"))
      .map((m) => m.name);

    let selectedModelName = GEMINI_MODEL_PRIORITY.find((candidate) => availableModels.includes(candidate));
    if (!selectedModelName && availableModels.length) {
      selectedModelName = availableModels[0];
    }
    if (!selectedModelName) {
      throw new Error("No Gemini models supporting generateContent are available.");
    }

    const genResp = await axios.post(
      `${GEMINI_API_BASE}/${selectedModelName}:generateContent`,
      { contents: [{ parts: [{ text: prompt }] }] },
      { params: { key: this._geminiApiKey }, timeout: 60000 }
    );

    const candidates = genResp.data.candidates || [];
    const text = (candidates[0]?.content?.parts || []).map((p) => p.text || "").join("").trim();
    if (!text) {
      throw new Error("Gemini returned an empty response.");
    }
    return text;
  }

  async _callNvidia(prompt) {
    const errors = [];

    for (const modelName of NVIDIA_CANDIDATE_MODELS) {
      try {
        const resp = await axios.post(
          `${NVIDIA_BASE_URL}/chat/completions`,
          {
            model: modelName,
            messages: [
              {
                role: "system",
                content:
                  "You are a senior technical evaluator. You must return only valid, " +
                  "well-formed JSON with no extra conversational text or markdown explanation.",
              },
              { role: "user", content: prompt },
            ],
            max_tokens: 4096,
            temperature: 0.2,
          },
          {
            headers: {
              Authorization: `Bearer ${this._nvidiaApiKey}`,
              "Content-Type": "application/json",
            },
            timeout: 60000,
          }
        );

        const choices = resp.data.choices || [];
        if (!choices.length) continue;

        const msg = choices[0].message || {};
        let text = (msg.content || "").trim();
        if (!text && msg.reasoning_content) {
          text = String(msg.reasoning_content).trim();
        }

        if (text) return text;
      } catch (exc) {
        const detail = exc.response ? JSON.stringify(exc.response.data) : exc.message;
        errors.push(`${modelName}: ${detail}`);
        continue;
      }
    }

    throw new Error(`NVIDIA candidates failed or returned empty: ${errors.join(" | ")}`);
  }

  /** Resilient router: try Gemini, then fall back to NVIDIA NIM models. */
  async generateText(prompt) {
    const errors = [];

    if (this._geminiApiKey) {
      try {
        return await this._callGemini(prompt);
      } catch (exc) {
        errors.push(`Gemini Error: ${exc.message}`);
      }
    }

    if (this._nvidiaApiKey) {
      try {
        return await this._callNvidia(prompt);
      } catch (exc) {
        errors.push(`NVIDIA Error: ${exc.message}`);
      }
    }

    if (errors.length) {
      throw new LLMEngineError(errors.join(" | "));
    }
    throw new LLMEngineError("No LLM provider is configured.");
  }

  static _stripCodeFences(text) {
    text = text.trim();
    if (text.startsWith("```")) {
      text = text.replace(/^`+|`+$/g, "").trim();
      if (text.toLowerCase().startsWith("json")) {
        text = text.slice(4);
      }
      text = text.trim();
    }
    return text;
  }

  static _isolateJsonObject(text) {
    const start = text.indexOf("{");
    const end = text.lastIndexOf("}");
    if (start === -1 || end === -1 || end < start) {
      throw new LLMEngineError(`No JSON object found in LLM response: ${text.slice(0, 150)}...`);
    }
    return text.slice(start, end + 1);
  }

  static _repairCommonJsonIssues(jsonStr) {
    jsonStr = jsonStr.replace(/\u201c|\u201d/g, '"');
    jsonStr = jsonStr.replace(/\u2018|\u2019/g, "'");
    jsonStr = jsonStr.replace(/,\s*([}\]])/g, "$1");
    jsonStr = jsonStr.replace(/\r/g, " ");
    return jsonStr;
  }

  static extractJson(rawText) {
    if (!rawText) {
      throw new LLMEngineError("Empty response from LLM provider.");
    }

    const text = LLMEngine._stripCodeFences(rawText);
    const jsonSlice = LLMEngine._isolateJsonObject(text);

    try {
      return JSON.parse(jsonSlice);
    } catch (exc) {
      const repaired = LLMEngine._repairCommonJsonIssues(jsonSlice);
      try {
        return JSON.parse(repaired);
      } catch (exc2) {
        throw new LLMEngineError(`Failed to parse JSON from LLM response: ${exc2.message}`);
      }
    }
  }

  /**
   * Runs the prompt, extracts JSON, and validates it against a zod schema.
   * @param {string} prompt
   * @param {import('zod').ZodSchema} schema
   */
  async generateStructured(prompt, schema) {
    const rawText = await this.generateText(prompt);
    const data = LLMEngine.extractJson(rawText);
    const result = schema.safeParse(data);
    if (!result.success) {
      throw new LLMEngineError(`LLM output failed schema validation: ${result.error.message}`);
    }
    return result.data;
  }
}

LLMEngine._instance = null;

module.exports = { LLMEngine, LLMEngineError };
